### PR TITLE
Add execution deadlines to worker util

### PR DESCRIPTION
**Pull request by eseliger**
Before this was an executor-only feature, but it should've been a worker native feature. Moved it over now :)

### CODE REVIEW

File: enterprise/cmd/executor/internal/worker/handler.go [hunk 1]
```diff
@@ -61,16 +61,6 @@ func (h *handler) PreDequeue(ctx context.Context) (dequeueable bool, extraDequeu
...
-	ctx, cancel := context.WithDeadline(ctx, time.Now().Add(h.options.MaximumRuntimePerJob))
-	defer cancel()
-
-	wrapError := func(err error, message string) error {
-		if errors.Is(err, context.DeadlineExceeded) {
-			err = errors.Errorf("job exceeded maximum execution time of %s", h.options.MaximumRuntimePerJob)
-		}
-
-		return errors.Wrap(err, message)
-	}
```
Comment (eseliger) [inline enterprise/cmd/executor/internal/worker/handler.go:73]: moving this up makes the errors look slightly different, the contextdeadlineexceeded part would now still be part of the error msg. Feelings about this?

File: internal/workerutil/worker.go [hunk 2]
```diff
@@ -315,8 +319,19 @@ func (w *Worker) handle(ctx, workerContext context.Context, record Record) (err
 	ctx, endOperation := w.options.Metrics.operations.handle.With(ctx, &err, observation.Args{})
 	defer endOperation(1, observation.Args{})
 
+	// If a maximum runtime is configured, set a deadline on the handle context.
+	if w.options.MaximumRuntimePerJob > 0 {
```
Comment (eseliger) [inline internal/workerutil/worker.go:323]: Wasn't sure if this is idiomatic go?

Comment (efritz) [reply internal/workerutil/worker.go:323]: We treat 0 as infinite timeout in other places.

File: internal/workerutil/worker.go [hunk 3]
```diff
@@ -315,8 +319,19 @@ func (w *Worker) handle(ctx, workerContext context.Context, record Record) (err
...
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithDeadline(ctx, time.Now().Add(w.options.MaximumRuntimePerJob))
+		defer cancel()
+	}
+
 	handleErr := w.handler.Handle(ctx, record)
 
+	if w.options.MaximumRuntimePerJob > 0 && errors.Is(handleErr, context.DeadlineExceeded) {
+		handleErr = errors.Wrap(handleErr, fmt.Sprintf("job exceeded maximum execution time of %s", w.options.MaximumRuntimePerJob))
```
Comment (efritz) [inline internal/workerutil/worker.go:332]: In the deleted code we didn't wrap, we just replaced it with a canned deadline exceeded error.

File: enterprise/cmd/executor/internal/worker/handler.go (hunk 1 above)
Comment (efritz) [reply enterprise/cmd/executor/internal/worker/handler.go:73]: See other comment

File: internal/workerutil/worker.go (hunk 3 above)
Comment (eseliger) [reply internal/workerutil/worker.go:332]: and then wrapped that error with the inner message, which we don't have at hand here anymore 🤔

File: enterprise/cmd/executor/internal/worker/handler.go (hunk 1 above)
Comment (efritz) [reply enterprise/cmd/executor/internal/worker/handler.go:73]: 😮‍💨 there no early return pls ignore me

**Review by efritz — APPROVED** [review]

**Pull request merged**